### ISSUE TITLE
ArgoCD very slow on inial load after upgrade to 2.3.3 from 2.2.5

### ISSUE THREAD

**Issue by bo0ts** [argoproj/argo-cd#9333]
**Describe the bug**

After the upgrade from 2.2.5 to 2.3.3 the ArgoCD webinterface takes a very long time to load. We have two ArgoCD instances with a small amount of apps (~50 per instance) and both exhibit this behavior. When looking at the network traffic in a developer console the main problem seems to be `main.c7ea22e999b3805bc676.js`.

I could reproduce the problem on Chrome and Firefox.

None of the pods running ArgoCD show signs of resource starvation.

**Screenshots**

![image](https://user-images.githubusercontent.com/76102/167384765-06a3ead3-f00d-4980-b850-1ca9720839c7.png)

**Version**

We deploy ArgoCD using the ArgoCD Operator on OKD 4.9. The ArgoCD 2.3.3 Update was triggered by the Operator Update to version 0.3.0.

```shell
argocd: v2.3.3+07ac038
  BuildDate: 2022-03-30T00:06:18Z
  GitCommit: 07ac038a8f97a93b401e824550f0505400a8c84e
  GitTreeState: clean
  GoVersion: go1.17.6
  Compiler: gc
  Platform: linux/amd64
```

**Comment by Maxcoder-net** [comment]
HI

Since when is this slowness found? Is it a daily slowness for AgroCD? What about other users , are they experiencing same slowness?

**Comment by bo0ts** [comment]
@Maxcoder-net Like I said, since the upgrade to 2.3.3 from 2.2.5. The problem appears when doing a full reload, but not while navigating in the app. It happens for every user on Firefox and Chrome (we haven't tested other browsers).

**Comment by viggin543** [comment]
Happened to me 2 !

**Comment by thnk2wn** [comment]
Maybe it's a different issue but I'm seeing the same behavior with `v2.2.2+03b17e0`. If I haven't hit the website in a bit it can take up to a minute to load. Once I'm there it's fast doing what I need.

**Comment by crenshaw-dev** [comment]
The differences in the timings is surprising. @bo0ts sees 9s, and @viggin543 sees 3min.

This is gonna be pretty difficult to diagnose. It involves the API server, potentially an ingress, the network between that and your client, the client itself... A packet trace and/or logs from each of those components would be a start.

**Comment by thnk2wn** [comment]
Appears to be main.js for myself as well, can try to look into logs...

![2022-06-21_15-18-02](https://user-images.githubusercontent.com/120529/174880787-c1cf2ae5-1045-456d-ae69-00a544ac06da.jpg)

**Comment by crenshaw-dev** [comment]
This is reproducible on ever refresh (or maybe hard refresh)? Do you happen to know if the API server is under heavy CPU and/or network load?

**Comment by thnk2wn** [comment]
> This is reproducible on every refresh (or maybe hard refresh)? Do you happen to know if the API server is under heavy CPU and/or network load?

I thought it was only when logged out but it does seem to be the same amount of time on a page refresh (normal or hard).

I didn't setup Argo and am not too familiar with its internals but if this helps:

![2022-06-21_15-29-12](https://user-images.githubusercontent.com/120529/174882752-f2f3d5f5-14ab-4c6a-bcea-293ca3dede46.jpg)

**Comment by thnk2wn** [comment]
A log file if it helps: [argocd-server-69fdcc9dc8-cjmkv.log](https://github.com/argoproj/argo-cd/files/8952318/argocd-server-69fdcc9dc8-cjmkv.log). Not sure what other component logs or steps would help.

**Comment by crenshaw-dev** [comment]
Low load. Looks like JS requests aren't logged, so that's not really gonna help us tell whether the API server is to blame. The fact that it's different in different locations makes me want to blame the network, but I don't want to jump to conclusions.

I think my next step would be running a packet trace on the client and on the API server. If the clocks are reasonably well synced, you can compare when the data packets arrive on the client vs. when the ACK packets make it back to the server to tell whether the network is to blame.

Thinking back to my web perf days, I guess it's also possible that a lot of packets are being dropped, forcing the packet size to stay super low.

**Comment by thnk2wn** [comment]
This was mostly fixed for me by turning on gzip compression (not sure why it wasn't on by default): https://github.com/argoproj/argo-cd/discussions/10238#discussioncomment-3942411

It can still take 6-9 seconds or so to initially load the page but much better than a minute.

**Comment by crenshaw-dev** [comment]
Yeah, that's still absurdly slow. But the fact that gzip helped is interesting.

**Comment by jlongo-encora** [comment]
Any updates on this one? I'm able to reproduce it on v2.4.14

**Comment by crenshaw-dev** [comment]
@jlongo-encora I think we need more details about the problem. So far, I think it's unclear whether Argo CD is misbehaving (sending stuff over the network too slowly) or if network conditions are the problem (or some combination of both).

One helpful piece of information would be seeing which assets are taking so long to transfer and at what rate they're being transferred.  We could compare transfer rate of, say, the main JS bundle to some other network response.

**Comment by jlongo-encora** [comment]
@crenshaw-dev 

![image](https://github.com/argoproj/argo-cd/assets/21343782/4d2c7f62-1751-4796-bcbc-f71f63ab356f)

**Comment by crenshaw-dev** [comment]
This is wild. The tiny extensions.js, a static asset, takes 9s. Two different userinfo responses have very different reaponse times.

Is this API server pod under a lot of load? Is it possible CPU throttling is monkeying with response times?

**Comment by jlongo-encora** [comment]
@crenshaw-dev we have ~350 applications

**Comment by crenshaw-dev** [comment]
Actually, extensions.js isn't quite static. So it could be a victim of throttling.

@jlongo-encora throttling won't be a direct function of number of applications. It happens server-side rather than on your browser. When a Kubernetes Pod uses more CPU than its configured limits, Kubernetes will "pause" CPU activity to avoid letting the Pod take too much time from other Pods on the node.

**Comment by jlongo-encora** [comment]
@crenshaw-dev ok thx for the explanation

**Comment by crenshaw-dev** [comment]
I think kubectl has the ability to show CPU usage. I always use Grafana, which my team set up to monitor our stuff. Unfortunately I don't know the details of that setup.

**Comment by thilinajayanath** [comment]
I had the same issue where a 600-ish KB `main.cxxx....js` file was taking over 50 seconds to load and it was fixed after I cleared the cache.
